docs/internal/bimonad.md

```diff
@@ -0,0 +1,353 @@
+# Bimonad tutorial
+
+Bifunctors have become a hot topic in scala since ZIO introduction of a second type parameter.
+It's a great way to manage business errors in your code, but it lacks suitable abstractions.
+The most comprehensive set I've seen lies [here](https://github.com/7mind/izumi/tree/develop/fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio).
+But it's built upon practical usage.
+So let's start building a more principled approach. 
+
+## Monads
+
+Let's talk about the basic definitions. We will use the basic category theory terminology, most of which can be found on [Wikipedia](https://en.wikipedia.org/wiki/Monad_(functional_programming)) and [nLab](https://ncatlab.org/nlab/show/monad).
+We will use the following definitions
+
+### Category 
+Starting from this section we would think of Category as some entity with some additional structure.
+The class mentioned below is something like a Set or Type. Its elements are objects of the category. Since we want the object class to be arbitrary big (bigger than classical Set could be) and we do not need equality\identity relation on it, we will say _Class_
+
+
+[Category](https://en.wikipedia.org/wiki/Category_(mathematics)) `C` is a Class called _objects_ together with indexed set family `Hom: (C, C) -> Set` and functions
+
+```
+id : ∀(a: C), Hom(a, a)
+
+∘ : ∀{a, b, c : C}, (Hom(b, c), Hom(a, b)) -> Hom(a, c) 
+```
+and following properties
+```
+left-id : ∀{a, b: C}, ∀(f: Hom(a, b)), id b ∘ f = f
+
+right-id : ∀{a, b: C}, ∀(f: Hom(a, b)), f ∘ id a = f
+
+associativity : ∀{a, b, c, d: C}, ∀(f: Hom(c, d)), ∀(g: Hom(b, c)), ∀(h: Hom(c, d)), ∀(h: Hom(a, b)), f ∘ (g ∘ h) = (f ∘ g) ∘ h
+```
+
+The most acknowledged category in a functional programming language is a `Set-like category of inhabitable types and pure total functions between them, so
+objects are types, and for types `A` and `B` `Hom(A, B)` is a set of pure functions `A => B`
+For scala, we will call it `Scala`
+
+Another form of a category of scala types is subtyping relations,
```

Comment:
```suggestion
Another sort of category that Scala types form is subtyping relations, i.e. we have a single element `Hom(A, B)` when `A <: B` and an empty Hom otherwise.
```